### ISSUE TITLE
hint about binding for optional @Input()

### ISSUE THREAD

**Issue by zoechi** [dart-lang/angular_analyzer_plugin#378]
![image](https://user-images.githubusercontent.com/405837/27291401-659e63b8-5510-11e7-966a-0f705a88325e.png)

```dart
@Directive(
  selector: '[virtualScroll]',
)
class VirtualScrollDirective  {
  bool _active;
  @Input()
  set virtualScroll(bool active) {
    if (active == _active) {
      return;
    }

    _active = active;
    if (_active) {
      _restoreState();
    } else {
      _persistState();
    }
  }
```

I think it shouldn't cause a hint/warning

**Comment by mk13** [comment]
This has been a pain point in dealing with input-bindings since there is a default, implied assumption that an empty-string exists within the value.

I guess we'll have to add some exceptions to ensure that this isn't raised for bool input types.

**Comment by MichaelRFairhurst** [comment]
This error is correct. You get empty string when using this binding like
you showed, not a boolean. If you check the angular_components lib, you'll
see they accept `dynamic` in the setter and treat `''` as `true`. You can
do the same to get the error to go away.

This really is more of an angular design bug; its an unexpected behavior,
and the workaround is not perfect either as type checking is turned off for
the input and `[mybool]="''"` counterintuitively enables the flag.

On Jun 19, 2017 1:02 PM, "Max Kim" <notifications@github.com> wrote:

This has been a pain point in dealing with input-bindings since there is a
default, implied assumption that an empty-string exists within the value.

I guess we'll have to add some exceptions to ensure that this isn't raised
for bool input types.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/dart-lang/angular_analyzer_plugin/issues/378#issuecomment-309557326>,
or mute the thread

**Comment by zoechi** [comment]
@MichaelRFairhurst I see. I didn't interpret the hint as type violation.
If `bool` had been mentioned in the message, it would be more clear.
Is there a way to improve the message. (instead of "is not a string" use "is of type bool")

**Comment by mk13** [comment]
Technically speaking, the `virtualScroll` can still be used a plain-attribute directive no? 
Only when it has an assigned value is it implicitly defined as using the @Input version of `virtualScroll`.

**Comment by MichaelRFairhurst** [comment]
What about

"using the attribute `virtualScroll` without brackets binds the attribute value `""`, a `String`, to the input `virtualScroll` on `VirtualScrollDirective`, which is expected to be of type `String`, but is actually of type `bool`."

**Comment by zoechi** [comment]
Perhaps it's just an issue because I had so many false hints and warnings in the past because of the early stage of the plugin. 
When I start trusting the hints more (recent updates got rid of most false hints already), I'll try harder to understand the messages before creating an issue and complaining about it ;-)
I see it's not too easy to fix/improve because of the somewhat confusing behavior of Angular.

@MichaelRFairhurst I think this is a good suggestion

> "using the attribute virtualScroll without brackets binds the attribute value ""

because it's not obvious that there is a binding happening at all, because the intention was just to get the directive added (the bound value is for special cases), and I think thats the root cause for the confusion.

**Comment by mk13** [comment]
> Perhaps it's just an issue because I had so many false hints and warnings in the past because of the early stage of the plugin.
When I start trusting the hints more (recent updates got rid of most false hints already), I'll try harder to understand the messages before creating an issue and complaining about it ;-)

In this stage of development, I think Mike and I would prefer it if you reported every suspicious issue you see :) This is just the rare case where the usage is not entirely clear either - but hopefully an error message with more clarity would help.

**Comment by zoechi** [comment]
> I think Mike and I would prefer it if you reported every suspicious issue 

@mk13 I have reported all issues I encountered so far. 
Until a few days ago I had dozens of false hints I couldn't get rid of, but these are now fixed since the last update a few days ago.
So now a new era begins :-) 
This one is a new one that I encountered while adding new code.
